File: LeetCode Problems/1 to 100/39. Combination Sum/39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(remaining, current, start):
            if remaining == 0:
                result.append(current[:])
                return
            if remaining < 0:
                return

            for i in range(start, len(candidates)):
                current.append(candidates[i])
                backtrack(remaining - candidates[i], current, i)  # Recurse with the same candidate allowed
                current.pop()

        result = []
        backtrack(target, [], 0)
        return result

File: LeetCode Problems/1 to 100/39. Combination Sum/39. CombinationSum.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        allCombinations(0, candidates, target, ans, new ArrayList<>());
        return ans;

    }

    private void allCombinations(int i, int[] candidates, int target, List<List<Integer>> ans,
            ArrayList<Integer> arrayList) {
        if (i == candidates.length) {
            if (target == 0) {
                ans.add(new ArrayList<>(arrayList));
            }

            return;
        }
        if (target == 0) {
            ans.add(new ArrayList<>(arrayList));
            return;
        }
        if (candidates[i] <= target) {
            arrayList.add(candidates[i]);
            allCombinations(i, candidates, target - candidates[i], ans, arrayList);
            arrayList.remove(arrayList.size() - 1);
        }
        allCombinations(i + 1, candidates, target, ans, arrayList);
    }
}
